airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py

```diff
@@ -0,0 +1,240 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+
+from enum import Enum
+from os import remove
+from time import sleep, time
+from typing import Any, Iterable, Mapping, Optional
+
+import requests
+from airbyte_cdk import AirbyteLogger
+from requests.exceptions import JSONDecodeError
+from source_shopify.utils import ApiTypeEnum
+from source_shopify.utils import ShopifyRateLimiter as limiter
+
+from .exceptions import ShopifyBulkExceptions
+from .query import ShopifyBulkQuery, ShopifyBulkTemplates
+from .record import END_OF_FILE, ShopifyBulkRecord
+from .tools import BulkTools
+
+
+class ShopifyBulkStatus(Enum):
+    CREATED = "CREATED"
+    COMPLETED = "COMPLETED"
+    RUNNING = "RUNNING"
+    CANCELED = "CANCELED"
+    FAILED = "FAILED"
+    TIMEOUT = "TIMEOUT"
+    ACCESS_DENIED = "ACCESS_DENIED"
+
+
+class ShopifyBulkJob:
+
+    """
+    Class to create, check, retrieve the result for Shopify GraphQL Bulk Jobs.
+    """
+
+    def __init__(self, session: requests.Session, logger: AirbyteLogger, query: ShopifyBulkQuery) -> None:
+        self.session = session
+        self.logger = logger
+        self.record_producer: ShopifyBulkRecord = ShopifyBulkRecord(query)
+        self.tools: BulkTools = BulkTools()
+
+    # 5Mb chunk size to save the file
+    retrieve_chunk_size = 1024 * 1024 * 5
+
+    # time between job status checks
+    job_check_interval_sec: int = 5
+
+    # max attempts for job creation
+    concurrent_max_retry = 10
+    # attempt counter
+    concurrent_attempt = 0
+    # attempt limit indicator
+    concurrent_max_attempt_reached = False
+    # sleep time per creation attempt
+    concurrent_interval_sec = 45
+
+    def get_errors_from_response(self, response: requests.Response) -> Iterable[Mapping[str, Any]]:
+        try:
+            return response.json().get("errors") or response.json().get("data", {}).get("bulkOperationRunQuery", {}).get("userErrors", [])
+        except (Exception, JSONDecodeError) as e:
+            raise ShopifyBulkExceptions.BulkJobBadResponse(
+                f"Couldn't check the `response` for `errors`, response: `{response.text}`. Trace: {repr(e)}."
+            )
+
+    def has_running_concurrent_job(self, response: requests.Response) -> bool:
+        """
+        When concurent BULK Job is already running for the same SHOP we receive:
+        Error example:
+        [
+            {
+                'field': None,
+                'message': 'A bulk query operation for this app and shop is already in progress: gid://shopify/BulkOperation/4039184154813.',
+            }
+        ]
+        """
+
+        concurent_job_pattern = "A bulk query operation for this app and shop is already in progress"
+        errors = self.get_errors_from_response(response)
+        if errors:
```

Comment:
The  `job_check_for_errors` does the additional checks and raises the per-defined exception,  including unexpected errors we haven't faced yet. This is the 3-step checks:
1) right after the  job  was created - we look for the job errors
2) while the job is `RUNNING` we check for errors/job check
3) if concurrent jobs are running we do the error check on every attempt

I could have missed something, feel free to suggest more places to cover with error handling.

Reply:
actually, looking through the code, I think the flow and error handling would be easier to understand if we passed around the errors instead of the response.

something like

```
    def job_create(self, request: requests.Request) -> Optional[requests.Response]:
        # create the server-side job
        response = self.session.send(request)
        # when the concurrent job takes place, we typically need to wait and retry, but no longer than 10 min.
        errors = self.get_errors_from_response(response)
        if not self.has_running_concurrent_job(errors):
            # process created job response
            return response if not self.job_check_for_errors(errors) else None
        else:
            return self.job_retry_on_concurrency(request)
```

this also removes the need to call `get_errors_from_response` multiple times for the same response

Reply:
Make sense to me, to  pass the `errors` iterable only, while checking for the rate-liimits in `get_errors_from_response`. Modified.